src/dynamic.ts

```diff
@@ -0,0 +1,78 @@
+import FixedThreadPool, {
+  FixedThreadPoolOptions,
+  WorkerWithMessageChannel
+} from './fixed'
+
+import { EventEmitter } from 'events'
+
+class MyEmitter extends EventEmitter {}
+
+export type DynamicThreadPoolOptions = FixedThreadPoolOptions
+
+/**
+ * A thread pool with a min/max number of threads, is possible to execute tasks in sync or async mode as you prefer.
+ *
+ * This thread pool will create new workers when the other ones are busy, until the max number of threads,
+ * when the max number of threads is reached, an event will be emitted, if you want to listen this event use the emitter method.
+ *
+ * @author [Alessandro Pio Ardizio](https://github.com/pioardi)
+ * @since 0.0.1
+ */
+export default class DynamicThreadPool<
+  Data = any,
+  Response = any
+> extends FixedThreadPool<Data, Response> {
+  public readonly emitter: MyEmitter
+
+  /**
+   * @param min Min number of threads that will be always active
+   * @param max Max number of threads that will be active
+   * @param filename A file path with implementation of `ThreadWorker` class, relative path is fine.
+   * @param opts An object with possible options for example `errorHandler`, `onlineHandler`. Default: `{ maxTasks: 1000 }`
+   */
+  public constructor (
+    public readonly min: number,
+    public readonly max: number,
+    public readonly filename: string,
+    public readonly opts: DynamicThreadPoolOptions = { maxTasks: 1000 }
+  ) {
+    super(min, filename, opts)
+
+    this.emitter = new MyEmitter()
+  }
+
+  protected _chooseWorker (): WorkerWithMessageChannel {
+    let worker: WorkerWithMessageChannel
+    for (const entry of this.tasks) {
+      if (entry[1] === 0) {
+        worker = entry[0]
+        break
+      }
+    }
+
+    if (worker) {
+      // a worker is free, use it
+      return worker
+    } else {
+      if (this.workers.length === this.max) {
+        this.emitter.emit('FullPool')
+        return super._chooseWorker()
+      }
+      // all workers are busy create a new worker
+      const worker = this._newWorker()
+      worker.port2.on('message', (message) => {
+        if (message.kill) {
+          worker.postMessage({ kill: 1 })
+          worker.terminate()
+          // clean workers from data structures
+          const workerIndex = this.workers.indexOf(worker)
+          this.workers.splice(workerIndex, 1)
+          this.tasks.delete(worker)
+        }
+      })
+      return worker
+    }
+  }
+}
+
+module.exports = DynamicThreadPool
```

Comment:
There's still something that I do not understand:

+ You have removed module.exports.ThreadWorker = ThreadWorker inside its class file;
+ You have kept module.exports = DynamicThreadPool and module.exports = FixedThreadPool inside their respective class files;
+ And there's module.exports = { FixedThreadPool, DynamicThreadPool, ThreadWorker } inside the index.ts file.

I think the legacy module export syntax is only needed in the index.ts file and can be removed elsewhere.

Reply:
I tried the library with a productive project before and after the rewrite to typescript
Without these `module.exports` for `DynamicThreadPool` and `FixedThreadPool` the user needs to use another import statement syntax as before
That's what I mean with a breaking API change
So I added these `module.exports` where necessary so the package exports the same _API_ as without the typescript rewrite
And yes we should remove these exports with #74 

After merging #77 it could be possible to release a last v1.3.0 or so. This way at least dependend projects using `1.x.x` benefit also from the typescript rewrite

Then we can build a `2.x.x` and if you want so we can pack any feature and optimizations into this as you want

Reply:
I'm perfectly fine the module.exports usage for backward compatibility. 
But I just do not understand why they are needed in all these places. 

Isn't `module.exports = { FixedThreadPool, DynamicThreadPool, ThreadWorker } `replacing all the others module.exports? 
Have `module.exports = DynamicThreadPool` a different semantic than `module.exports = { DynamicThreadPool }`? 
I'm not very familiar with that legacy syntax and its semantic.

Reply:
https://stackoverflow.com/a/40295288/6897682

As I said, for v2 we will remove all these `module.exports` and just relay on `export / export default`

Reply:
That reads, are you sure then that you can remove safely `module.exports.ThreadWorker = ThreadWorker` from its class file? 

Thanks.

Reply:
Yes, because it's not `module.exports = ThreadWorker` but `module.exports.ThreadWorker = ThreadWorker`

And this is the equivalent of `export class ThreadWorker ...`